EZP-25627 provide default layout template allowing show sf toolbar

crevillo opened the pull request:
Here's another try to solve this. In this case i create a new `pagelayout.html.twig` template and a new param in `templating.yml` file. i called it `ezpublish.pagelayout`. Better naming suggestions are very welcomed :). 

The `full.html.twig`extends that new `viewBaseLayout`.
One concern about this is what if the user tries something like 
```twig
 {{ render( controller( "ez_content:viewAction", {'contentId': contentId, 'viewType': 'full', 'noLayout': 1} ) ) }}
```
In this case, shouldn't we modify full template to check if ```noLayout``` is true or false?

ping @andrerom @emodric @bdunogier

crevillo commented:
build green now.

emodric commented:
I'd rather not overwrite `viewbasePagelayout` variable in templates, but create a new one (called `pagelayout`). That way, if someone uses `noLayout` for full view they can still extend the `viewbase_layout.html.twig` with:

`{% extends noLayout ? viewbaseLayout ? pagelayout %}`

`ezpublish.pagelayout` should be siteaccess aware.

crevillo commented:
@emodric: you mean something like https://github.com/ezsystems/ezpublish-kernel/pull/1626/commits/55ec1eeb483ce94a453f762b377eab5313bb92fb ?
siteaccess part will come later.

emodric commented:
Exactly, only it should be `noLayout == true` ;) (If NO layout, use `viewbaseLayout` )

crevillo commented:
True that. I Will change it. Meeting now.

On Tue, 29 Mar 2016 12:13 PM Edi Modrić, <notifications@github.com> wrote:

> Exactly, only it should be noLayout == true ;) (If NO layout, use
> viewbaseLayout )
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/ezsystems/ezpublish-kernel/pull/1626#issuecomment-202820393>
>

crevillo commented:
@emodric, @bdunogier couldn't we make viewbase_layout siteaccess aware too?

emodric commented:
I guess we could. Is there a need though? It's a fallback empty template after all.

crevillo commented:
probably there's no need, but it won't hurt either. :). Someone could want to add different container per siteaccess...

crevillo commented:
now `pagelayout` is siteaccess aware. 
ping @emodric @bdunogier

emodric commented:
Looks good :) Although, maybe pagelayout could be injected directly with setter method and dynamic setting (`$pagelayout$`)

crevillo commented:
i think i need to study a little more for that ;)

crevillo commented:
re-ping @emodric is that what you meant? https://github.com/ezsystems/ezpublish-kernel/pull/1626/commits/45482d7679612fd497dbd16d64515a53910a5cc3

emodric commented:
@crevillo Yep, that's it!

crevillo commented:
cool. is it a +1 then? maybe @andrerom @bdunogier could add their pov and if they like i could squash all this.

emodric commented:
I can't test, I'm on vacation with limited access to my machine, but from the looks of it, it's a +1 :)

crevillo commented:
thanks @emodric for your time. didn't know you were on vacation. enjoy!.

andrerom commented:
@crevillo Looks good, but maybe we should make sure the pagelayout validates?
see: https://gist.github.com/cmalven/1885287

- I haven't checked where you might get the `lang` from
- As for `utf-8` we might be able to take that from somewhere but unsure if needed
- Title that can maybe be a block with default value, there is already convention for this right?

crevillo commented:
@andrerom, ok. we can try. note actually, the provided templates just produces the Welcome page without any html or body tag at all. :)

emodric commented:
About the `lang`, we use `<html lang="{{ app.request.locale }}">`

About `utf-8`, is there really need for parametrizing it, as opposed to hardcoding it with `<meta charset="utf-8">` ? In this day and age? :)

About the title, we usually use site name (taken from config with `{% set site_name = ezpublish.configResolver.getParameter( 'site_settings.site_name', 'some_namespace' ) %}`).

crevillo commented:
@emodric, @andrerom 
* for the lang, it looks {{ app.request.locale }} is not enough as this will produce something like `en_GB` instead of `en` or `en-GB`. I took this second approach and replaced the `_` by `-`. Ref: https://tools.ietf.org/html/rfc5646#section-2.1, also, you try it in w3c validator and `en_GB` will throw an error `Error: Bad value en_GB for attribute lang on element html: The language subtag en_gb is not a valid language subtag.`
* for the meta charset. I've just added it hardcoded as @emodric suggest. it looks in the past symfony allowed define charset as a framework parameter, but this is not the case right now. Anyway, meta charset presence is no necessary to pass the validator. 
* for the title, i've just copied what we have in ezplatform-demo repo https://github.com/ezsystems/ezplatform-demo/blob/master/app/Resources/views/pagelayout.html.twig#L4-L7. This has a side effect, that, in a eZ Platform install the only provided content is called `eZ Platform` this will produce a `eZ Platform - eZ Platform` title for the welcome page. 

Let me know your thoughts on this change and i'll squash everything if you want.

emodric commented:
You're right about language code :+1: `replace` is perfectly fine with me, btw :)

As for title, as @andrerom suggested, maybe it would be okay to wrap in a Twig block, so it can be overriden easily?

crevillo commented:
@emodric, yep, also with that way we can avoid a bit of twig there. i re-ping @bdunogier as i think he has something more to say before we could merge this.

bdunogier commented:
Lots of discussions there. I'll have a look again after sf live (tomorrow).

In the meanwhile, ping @kevindejour & @dspe ;)

andrerom commented:
Looks good! As I hinted I think hardcding utf8 is fine.

The convention for title I had in mind, was the one in old demobundle: https://github.com/ezsystems/DemoBundle/blob/master/Resources/views/page_head.html.twig

So what if we do something like this as the minimal layout:
```jinja
<!doctype html>
<html lang="{{ app.request.locale|replace('_', '-') }}">
<head>
    <meta charset="utf-8">
    {% if content is defined and title is not defined %}
        {% set title = ez_content_name( content ) %}
    {% endif %}
    <title>{{ title|default( 'Home' ) }}</title>
    <meta name="generator" content="eZ Platform"/>
    {# adding canonical url for all defined content #}
    {% if content is defined and content.contentInfo.mainLocationId %}
        <link rel="canonical" href="{{ path( 'ez_urlalias', {'locationId': content.contentInfo.mainLocationId} ) }}" />
    {% endif %}
</head>
<body>
{% block content %}{% endblock %}
</body>
</html>
```

crevillo commented:
i also like @emodric propopsal. i was thinking in pros and contras. 
* following @emodric approach: we'll need to a ```{%block title%} ... {%endblock%]``` in every template to change the titles, won't we?
* following @andrerom 's one: for templates that are not meant to render a content (maybe a custom controller that loads lot of contents not all under the same folder) we'll need a {% set title='whatever' %} in those templates, but for the content ones, title will be shown without the need of any change... right?

At the end, it looks less work for the final developers what @andrerom suggests... so, @emodric ok to that?

about adding the canonical, i'm fine with it. doing.

andrerom commented:
> Edit2: Following @emodric approache to title, how about adding blocks for javascripts and styles? i mean

If this is convention already in other demo design templates, then why not, otherwise I would personally have a bigger need to actually add a global stylesheet and js file, which maybe highlights the fact that we are maybe overdoing this if the goal is to just have a temporary one, if we change the goal to have a fully customizable one however, then we have work to do :P

emodric commented:
I have no preference for adding JS and CSS blocks as I've never used any built in pagelayout :)

If the purpose of this pagelayout is only to enable Symfony toolbar, then it's not needed I guess. We're adding a title just because it's required by HTML spec.

As for title, it indeed may be better using a variable.

crevillo commented:
What @emodric says is correct. so, i'm gonna add last André's proposal and we could discuss about this in the future.

crevillo commented:
Added last @andrerom 's comments and squashed everything. Let's wait for @bdunogier. 
About the idea for javascript and sytles block: Obviously this is not legacy anymore, but when i started with eZ 5 one of the things i tried, probably wrongly, is to try to know which things could be similar in some way. actually, one of the first things i did was searching for a pagelayout :). 

so, yep, probably at the end nobody will use this template, but at least it can be a start point if someone wants to give it a try. and if so, i think the dev will be happier if they can customize things without touching / overriding lot of templates :)

emodric commented:
Exactly! The biggest point of this PR was (when we have a chance), to allow for pagelayout to be *configured* properly.

bdunogier commented on eZ/Bundle/EzPublishCoreBundle/DependencyInjection/Configuration/Parser/Common.php:
```diff
@@ -93,6 +93,10 @@ public function addSemanticConfig(NodeBuilder $nodeBuilder)
             ->end()
+            ->scalarNode('pagelayout')
+                ->info('The default layout to use')
+                ->example('EzPublishCoreBundle::pagelayout.html.twig')
```
Would a real life example be more instructive ?

crevillo replied:
yep, why not. gonna change it.

dspe commented:
That's a really nice WIP :+1: But regarding the default pagelayout.html.twig why not giving it more flexibility and add some js blocks, css blocks etc ?

crevillo commented:
@dspe That's something we already commented here, but as the goal of this was only show the toolbar, we decided to postpone that for a near future. Maybe we should open a jira or something and start from it.

andrerom commented:
And we have feature freeze on monday for 16.04, so if possible what can we agree on for this release train?

crevillo commented:
to me, this could be merged as it now. i think we just miss @bdunogier approval. :)

emodric commented:
Looks good to me too 👍

andrerom commented:
+1 from my side

@bdunogier your call :)

bdunogier commented:
My only "concern" is that we are introducing a setting we will have to maintain, while we are apparently not sure what the next improvement will be. We also have a lack of connection between the app's pagelayout and the default platform

But this setting has a clear name, can I think easily be moved forward while we enhance that layer, and we still have the option to deprecate/migrate it if we need.

emodric commented:
Why is creating another setting such an issue? Yes, you will need to maintain it, but you're a vendor, it's your job. This is something that should've been here from start, from 5.0. It's obvious that this is something that people use and need so much that I'm fairly certain each of them have it one way or the other already custom implemented.

> We also have a lack of connection between the app's pagelayout and the default platform ones, that might be confusing, and not what we want in the end.

I have no idea what this even means :) You can now use this in demo design or a customer project to parametrize pagelayout based on siteaccess, without ever touching full views. Just like you could before.

bdunogier commented:
> I have no idea what this even means :) You can now use this in demo design or a customer project to parametrize pagelayout based on siteaccess, without ever touching full views. Just like you could before.

I got sidetracked by other considerations. So much for being on vacation and focusing on gardening ;-)

Here we go, thank you very much @crevillo, as well as the others for the reviews and discussions. Let's follow-up :-)

bdunogier commented:
Integration to demo @ https://github.com/ezsystems/ezplatform-demo/pull/18
